src/mpi/coll/allreduce/allreduce_intra_recursive_multiplying.c

```diff
@@ -0,0 +1,202 @@
+                                            void *recvbuf,
+                                            MPI_Aint count,
+                                            MPI_Datatype datatype,
+                                            MPI_Op op,
+                                            MPIR_Comm * comm_ptr, 
+                                            const int k,
+                                            MPIR_Errflag_t errflag)
+{
+    int mpi_errno = MPI_SUCCESS;
+    /* Ensure the op is commutative */
+    MPIR_Assert(MPIR_Op_is_commutative(op));
+
+    int comm_size, rank, virt_rank;
+    comm_size = comm_ptr->local_size;
+    rank = comm_ptr->rank;
+    virt_rank = rank;
+    
+    /* get nearest power-of-two less than or equal to comm_size */
+    int power = (int) (log(comm_size) / log(k));
+    int pofk = (int) lround(pow(k, power));
+
+    MPIR_CHKLMEM_DECL(2);
+    void *tmp_buf;
+
+    /* need to allocate temporary buffer to store incoming data */
+    MPI_Aint true_extent, true_lb, extent;
+    MPIR_Type_get_true_extent_impl(datatype, &true_lb, &true_extent);
+    MPIR_Datatype_get_extent_macro(datatype, extent);
+    MPI_Aint single_node_data_size = extent * count - (extent - true_extent);
+    
+    MPIR_CHKLMEM_MALLOC(tmp_buf, void *, (k - 1) * count * single_node_data_size, mpi_errno,
+                        "temporary buffer", MPL_MEM_BUFFER);
+    
+    /* adjust for potential negative lower bound in datatype */
+    tmp_buf = (void *) ((char *) tmp_buf - true_lb);
+
+
+    /* copy local data into recvbuf */
+    if (sendbuf != MPI_IN_PLACE) {
+        mpi_errno = MPIR_Localcopy(sendbuf, count, datatype, recvbuf, count, datatype);
+        MPIR_ERR_CHECK(mpi_errno);
+    }
+
+    /* PRE-STAGE: ranks over pofk collapse their data into the others
+     * using reverse round-robin */
+    if (pofk < comm_size) {
+        if (rank >= pofk) {
+            /* This process is outside the main algorithm so send data */
+            int pre_dst = pofk - (rank % pofk) - 1;
+            /* This is follower so send data */
+            mpi_errno = MPIC_Send(recvbuf, count, datatype,
+                    pre_dst, MPIR_ALLREDUCE_TAG, comm_ptr, errflag);
+            MPIR_ERR_CHECK(mpi_errno);
+            /* Set virtual rank so this rank is not used in main stage */
+            virt_rank = -1;
+        } else {
+            int switch_bufs = 1;
+            /* Receive data from all those greater than pofk */
+            for (int pre_src = 2 * pofk - (rank % pofk) - 1; pre_src < comm_size; pre_src += pofk) {
+                if(switch_bufs) {
+                    mpi_errno = MPIC_Recv(tmp_buf, count, datatype, pre_src,
+                            MPIR_ALLREDUCE_TAG, comm_ptr, MPI_STATUS_IGNORE);
+                    MPIR_ERR_CHECK(mpi_errno);
+                    mpi_errno = MPIR_Reduce_local(recvbuf, tmp_buf, count, datatype, op);
+                    MPIR_ERR_CHECK(mpi_errno);
+                    switch_bufs = 0;
```

Comment:
You allocated `k-1` buffers so why not `MPIC_Irecv` all of them then local reduce all of them? That would reduce latency of receiving one by one.

Reply:
Good idea, fixed